Testing: Add failure state in testing context

Sgitario opened the pull request:
Sometimes it's necessary to know whether the tests failed (at test case level or even in the before stages). For example: if we use one resource via the `@QuarkusTestResource` annotation that is starting a service and we don't want to stop this service when the test fails. 

At the moment, we are not propagating when a test fails. 

This pull request addresses this issue by adding this information in:
- QuarkusTestContext (for QuarkusTest*Callbacks)
- Context (for QuarkusTestResources)

To be precise, I've added the following methods in the above contexts:

```java
/**
         * Get the test failure result if the test failed. When no test have failed, it will return null.
         *
         * @return the failure result that is thrown during either `BeforeAll`, `BeforeEach`, test method, `AfterAll` or
         *         `AfterEach` phases. 
         */
        TestStatus getTestStatus();
```

In the future, we could add even more information like the number of tests that have failed and/or the test profiles.

Fix https://github.com/quarkusio/quarkus/issues/25749

geoand commented:
Hm... 

I understand the logic of this, but don't these only work if called after the tests have finished?

Sgitario commented:
> Hm...
> 
> I understand the logic of this, but don't these only work if called after the tests have finished?

Correct. When using the after each or after all callbacks, the usage is natural. 
In the test resource, you would need to keep the context instance and check it in the `stop` method.

geoand commented:
This should be very very clear in the javadoc

metacosm reviewed (commented):
I need to give it a try but it looks good to me…

Sgitario commented on test-framework/common/src/main/java/io/quarkus/test/common/TestFailureResult.java:
```diff
@@ -0,0 +1,25 @@
+package io.quarkus.test.common;
+
+public class TestFailureResult {
```
I like the idea. @geoand do you agree with renaming the `TestFailureResult` to `TestExecutionStatus`?

geoand replied:
Yup, I like the idea

Sgitario commented:
PR updated with conflicts resolved. cc @metacosm @gsmet

gsmet commented:
I like it better but I prefer if we get @geoand to review and merge it. Let's wait for him to come back from the dead :).

geoand commented:
I likely won't be back this week

geoand reviewed (commented):
Seems reasonable, but the whole point of this was to support a use case that @metacosm has, so IMO we need him to try it out first

gsmet commented:
@metacosm you around to test that one?

metacosm commented:
@gsmet I need to check with the new changes, I was OK with the initial version.

metacosm commented:
As long as the test result is available from the context so that a test resource can figure out if tests have failed at some point, that's good enough for me 😄

geoand commented:
So should we merge this one?

Sgitario commented:
> So should we merge this one?

Yes, please. I've just rebased the pull request just in case.

geoand commented:
Great, thanks!

Sgitario committed:
Testing: Add failure state in testing context

Sometimes it's necessary to know whether the tests failed (at test case level or even in the before stages). For example: if we use one resource via the `@QuarkusTestResource` annotation that is starting a service and we don't want to stop this service when the test fails.

At the moment, we are not propagating when a test fails.

This pull request addresses this issue by adding this information in:
- QuarkusTestContext (for QuarkusTest*Callbacks)
- Context (for QuarkusTestResources)

To be precise, I've added the following methods in the above contexts:

```java
/**
         * Get the test failure result if the test failed. When no test have failed, it will return null.
         *
         * @return the failure result that is thrown during either `BeforeAll`, `BeforeEach`, test method, `AfterAll` or
         *         `AfterEach` phases.
         */
        TestStatus getTestStatus();
```

In the future, we could add even more information like the number of tests that have failed and/or the test profiles.

Fix https://github.com/quarkusio/quarkus/issues/25749

gsmet commented:
@geoand you marked this one as waiting for ci so I suppose it was good on your side? CI failures are just flakes.
